vllm/model_executor/layers/linear.py

```diff
@@ -538,6 +635,109 @@ def __init__(self,
                          params_dtype=params_dtype,
                          quant_config=quant_config)
 
+    def _get_shard_offset_mapping(self, loaded_shard_id: str):
+        shard_offset_mapping = {
+            "q": 0,
+            "k": self.num_heads * self.head_size,
+            "v": (self.num_heads + self.num_kv_heads) * self.head_size,
+            "total": (self.num_heads + 2 * self.num_kv_heads) * self.head_size
+        }
+        return shard_offset_mapping.get(loaded_shard_id)
+
+    def _get_shard_size_mapping(self, loaded_shard_id: str):
+        shard_size_mapping = {
+            "q": self.num_heads * self.head_size,
+            "k": self.num_kv_heads * self.head_size,
+            "v": self.num_kv_heads * self.head_size,
+        }
+        return shard_size_mapping.get(loaded_shard_id)
+
+    def _default_loading(self, param: vLLMParameter, param_data: torch.Tensor,
+                         loaded_weight: torch.Tensor, loaded_shard_id: str):
+
+        tp_rank = get_tensor_model_parallel_rank()
+        shard_offset = self._get_shard_offset_mapping(loaded_shard_id)
+        shard_size = self._get_shard_size_mapping(loaded_shard_id)
+
+        # TODO: Define this relationship more clearly
+        if param.is_packed and param.output_dim == param.packed_dim:
+            shard_size, shard_offset = param.adjust_packed_shard(
+                shard_offset=shard_offset, shard_size=shard_size)
+
+        param_data = param_data.narrow(param.output_dim, shard_offset,
+                                       shard_size)
+        if loaded_shard_id == "q":
+            shard_id = tp_rank
+        else:
+            shard_id = tp_rank // self.num_kv_head_replicas
+
+        loaded_weight.narrow(param.output_dim, shard_id * shard_size,
+                             shard_size)
+
+        return param_data, loaded_weight
+
+    def _load_no_shard_id(self, param: vLLMParameter, loaded_weight):
+        shard_offsets = [
```

Comment:
can we re-use `_get_shard_offset_mapping` and `_get_shard_size_mapping` here?

Reply:
Reuse from where?

Reply:
slightly different mappings